### ISSUE TITLE
Using Parse Server in production

### ISSUE THREAD

**Issue by zeluspudding** [ParsePlatform/parse-server#1106]
So this isn't really an issue but a question who's answer is probably good to know for a lot of us. Stemming from the implications of #1057 I'm wondering if ParseServer is ready for thin (or even fat for that matter) client applications. If this is not the appropriate place to answer this question, please, by all means reply to this (not my) [post](https://www.reddit.com/r/iOSProgramming/comments/491ivu/anyone_currently_using_parse_server_in_production/)

**Comment by tobernguyen** [comment]
I'm. But honestly, it is really unstable. I have had to switch back to Parse.com 3 times. So many unknown problems happened and I can't even reproduce it. My setup:

- Hosted mongoDB at mLab: Shared Cluster
- Two VPS running Parse Server
- One VPS for load balancing to 2 above server
- Background job processing with Kue and Redis
- NewRelic for performance measurement and Airbrake for exception notification.

**Comment by FridayDevGroup** [comment]
+1. Would like to hear some feedback from a reliability standpoint. These guys have been working very hard on it. Every time I check out this git there's a commit made just a few hours ago, and with the added features at this point parse-server has almost superseded it's predecessor.

**Comment by bohemima** [comment]
TL;DR - Wait with parse-server for at least 2-3 months, let it mature first.


I quickly jumped on the train here, been following pretty much all commits, issues, PR's, etc.

The very first thing you'll notice is how easy it is now to run and debug cloud code locally, this is one of the major improvements we get now.

Stability - I'd stay put for a while, there is no rush yet. You definitely want to migrate your database before the end of April, as the migration docs hints about lower priority for their own mongodb instances.

Following the migration docs, the next "important" date is July 28th, where you are supposed to switch to your own hosted parse-server. The migration docs are pretty spot on and very well written. Read it carefully.

**Comment by omairvaiyani** [comment]
We've switched to parse-server on production, but mainly because our platform is still in beta ~500 daily users. It's stable enough most of the time, but there seems to be a disconnect between the level of activity on parse-server vs. parse-sdk-js; we often find issues with the client-side code arising after updating parse-server. Overall, the issues have been balanced by the ability to test cloud code locally. The lack of timeouts is a massive bonus.

**Comment by davimacedo** [comment]
In back4app we have now about 500 apps running in production with Parse Server running in our infra-structure. Honestly most part of them are very simple applications. At this moment, I think that Parse Server is not 100% prepared for existing complex applications. The complex ones still suffer in the migration process mainly with features related to Jobs, Push Notifications and Cloud Code Hosting. BUT I really believe that in a little time Parse Server will be the best framework for API and backend development. We have just to keep working hard and focused to make it real: app developers sending bugs/issues, parse server developers improving the framework and parse hosting services improving tools to make it easier for developers.

**Comment by flovilmart** [comment]
what @bohemima said! There is no rush to migrate to parse-server. In the meantime, before July 28th, it's maybe a great time to train your team in deployments, shell scripts, collecting logs from cloud providers, find the provider that best suits you etc...

Use that time to have a good understanding of what you're leaping into, from a fully managed backend to a backend that you manage.

Read on best practices on deploying express apps, do testing on small apps, get familiar with the code base, the inner workings etc... As the code is open-source this doesn't mean you'll ever have full time engineers to fix/patch the problems you face. 

Post issues, pull requests, that's one of the best ways to get familiarized with the source.
In the end, even if you deploy to back4app or anywhere else, you'll need to understand why sometimes things don't work as expected. 

Keep in mind, that this code, as being open-source is provided as-is. There is nothing forcing the contributors to work on the bugs you discover.
The core contributors are mostly here to ensure feature parity is achieved in a reasonable time frame and make sure the quality of the code in the pull requests match the requirements for such project.

**Comment by flovilmart** [comment]
@tobernguyen that would be nice if you managed to repro those problems, that would be valuable information for many users

**Comment by zeluspudding** [comment]
These comments help frame what should be a reasonable schedule for getting parse server into production... I know, at least for me, it will be somewhat longer than originally desired. 
That's OK.
The team that's come around the project thus far has done wonderful work! Thank you all for your input - it really does help.

**Comment by corbanb** [comment]
@tobernguyen would you be able to share more on how you setup the background jobs with kue and redis?

**Comment by carmenlau** [comment]
Would like to ask you guys after you migrate to the new parse server, will you keep your old app running with parse.com cloud code? During development, I found some backward compatibility problems. Like parse.com and parse server are using different fields for ACL. The old app are not able to get the images from new s3 buckets...etc.

**Comment by FridaySG** [comment]
@carmenlau You should document those problems and open up some issues with logs if you can please.

**Comment by alexrmacleod** [comment]
@omairvaiyani how did u get 500 daily users in beta. i cant even get 100 in my latest production release

**Comment by omairvaiyani** [comment]
@alexrmacleod i don't think that is relevant to parse server but for your curiosity, it's an educational platform and we have a wide reach from our previous product.

**Comment by mitchellporter** [comment]
> You definitely want to migrate your database before the end of April, as the migration docs hints about lower priority for their own mongodb instances.

This is what worries me. If Parse Server is not actually ready for production apps by the end of April, then  it sounds like you're kind of screwed. If you have a successful app in production that is currently hosted with Parse.com, can you actually migrate to Parse Server by end of April and provide a stable experience for your users, or will your users suddenly be experiencing issues now?

If it's not stable/production ready by end of April and you really want to keep using Parse's tools, then you have no choice but to stay on Parse.com for the time being. What will this mean for your app, and your user's experience? Will an app with 10k users start slowing down? What about 50k, 100k+ users?

Please note that none of this is meant as criticism, I'm simply trying to have an open discussion and figure out where we all are on this issue. The parse team has done an amazing job of open sourcing this and I truly hope that we can build a strong community around this.

My problem is that I'm an iOS dev, not a backend dev. As much as there is lots of documentation on how to migrate, most of us won't have the skills yet to handle all of these potential issues that might occur when moving to Parse Server.

This situation might be fine for side projects or personal apps, but what about those of us running a business? What about those of us trying to launch apps that will be featured in the store and receive 100k+ downloads? What about apps built on Parse that already have 100k+ users?

The last thing I want to do is stop using Parse. It would be a shame because I'm already familiar with the iOS SDK and there's already so much built for us.

I'm just trying to take a pulse of the situation and see where everyone is at with this. At this point I'm excited about the potential that Parse Server has, but at the same time I'm worried that it might not be ready in time before they start winding things down over there.

**Comment by drew-gross** [comment]
@mitchellporter you can keep your app server on Parse.com after you've migrated your database. We recommend migrating your database by the end of April, but migrating your application logic to Parse Server can wait a few more months. If you have migrated your database by April 28th, we won't deprioritise your application server.

**Comment by mitchellporter** [comment]
@drew-gross Thanks for the clarification. For anyone else that's curious: https://github.com/ParsePlatform/parse-server/wiki/Migrating-an-Existing-Parse-App

**Comment by carmenlau** [comment]
@tobernguyen You said you had switched back to parse.com 3 time so far. Would like to know how do you do the switch? In client side OR server proxy to http://api.parse.com/? Want to find a way so we can do it safety. Thanks!:)

**Comment by hramos** [comment]
Very interesting thread, and thank you all who have started using Parse Server and contributed to its development. 

I am closing this out as we're trying to cut down on the number of non-issues. This will help us focus on the issues that are actively preventing people from migrating their production apps to Parse Server.

As mentioned earlier in the thread, the April 28 date is only about migrating your database. Calls to `api.parse.com` will continue working, and Parse will just route calls to your database instead of Parse's. 

You still have until July 28 to start using Parse Server. This will give you roughly six months to get your active users to update their client apps to the newer version that points to your self-hosted Parse Server before `api.parse.com` stops handling requests.
